f/connectors/alerts/alerts_gcs.py

```diff
@@ -124,221 +131,251 @@ def _get_tif_rel_filepath(blob_name, local_file_path, territory_id):
     """
-    Calculate the SHA-256 checksum of a file to ensure that `process_files` asset processes only new versions of `alerts_metadata.csv` (filename).
+    logger.info(f"Converting TIF files: {tiff_files}")
+    for tiff_file in tiff_files:
+        tiff_file_path = Path(tiff_file)
+        jpeg_file_path = tiff_file_path.with_suffix(".jpg")
+        jpeg_file = jpeg_file_path.name
+
+        # If the jpeg file already exists, skip it
+        if jpeg_file_path.exists():
+            logger.info(f"JPEG file already exists: {jpeg_file}")
+            continue
+
+        logger.info(f"Converting TIFF file to JPEG: {jpeg_file}")
+
+        # Ensure the file exists in the local directory
+        if tiff_file_path.exists():
+            try:
+                with Image.open(tiff_file_path) as img:
+                    # Save the image in the same location in the datalake as the tiff
+                    img.save(jpeg_file_path, "JPEG")
+            except Exception as e:
+                logger.error(
+                    f"TIFF image can not be opened, potentially empty: {str(e)}"
+                )
+
+    logger.info("Successfully converted TIFF files to JPEG.")
 
-    This function reads the file in chunks to avoid loading the entire file into memory, which is particularly useful for large files.
+
+def _generate_uuid(row):
+    """
+    Generate a unique UUID for a row using SHA256 hashing.
 
     Parameters
     ----------
-    filename : str
-        The path to the file whose checksum needs to be calculated.
-    bufsize : int (optional)
-        The size of the buffer used for reading the file in chunks. Default is 128 KB.
+    row : pd.Series
+        A pandas Series representing a row in the DataFrame.
 
     Returns
     -------
     str
-        The hexadecimal representation of the SHA-256 checksum of the file.
+        A unique UUID for the row.
     """
-    h = hashlib.sha256()
-    buffer = bytearray(bufsize)
-    buffer_view = memoryview(buffer)
-    with open(filename, "rb", buffering=0) as f:
-        while True:
-            n = f.readinto(buffer_view)
-            if not n:
-                break
-            h.update(buffer_view[:n])
-    return h.hexdigest()
-
-
-def process_files(path, files, territory_id, alerts_metadata_filename):
-    """Convert various `files`, specifically GeoJSON, TIFF, and CSV files.
-
-    1. Converts any TIFF file to JPEG format (but also keeping the TIFF file)
-    2. Collects the names of any .geojson file in path and returns it.
-    3. Filters the `alerts_metadata_filename` CSV rows based on the specified territory ID,
-      and generates a unique metadata UUID for the filtered CSV records, returning them.
+
+    def sha256sum(data):
+        h = hashlib.sha256()
+        h.update(data.encode("utf-8"))
+        return h.hexdigest()
+
+    row_data = json.dumps(row.to_dict(), sort_keys=True)
+    row_hash = sha256sum(row_data)
+    return str(uuid.uuid5(uuid.NAMESPACE_DNS, row_hash))
```

Comment:
I have some doubts about this new `_generate_uuid` function.

However before I get into them, can you explain (ideally in some code comments and point me to them) what the point of generating a UUID is for?  I see the PR description refer to avoiding duplicates but help me understand the bigger picture please.

That way I can be sure to give feedback that's relevant and not misguided.

Reply:
https://github.com/ConservationMetrics/gc-scripts-hub/blob/822d2ea52ef11baaad584484b3b0ec181d2522d0/f/connectors/alerts/alerts_gcs.py#L410 creates a PRIMARY_ID column `_id`. Previously, there were no PRIMARY KEY columns and hence alerts__metadata rows were being duplicated every time this script is run.

https://github.com/ConservationMetrics/gc-scripts-hub/blob/822d2ea52ef11baaad584484b3b0ec181d2522d0/f/connectors/alerts/alerts_gcs.py#L632 sets the value of `_id` as the row's UUID. And the reason we generate a UUID to begin with is because we don't get any unique identifier from the `alerts_history.csv` file otherwise.

Reply:
Thanks for explaining.

There is no _great_ solution here, every solution I think of is prone to edge cases, such as what datatype pandas uses to represent a column, or the order of the columns, etc.

My first suggestion is to consider whether you really want to consider all columns of the row, or only one or two "index columns." I'd think that [territory_id, month,  year] might be enough?  Or [territory_id, month,  year, alert_source, description_alerts].  That way if for whatever reason there was a change in some other field like "confidence" (only), we don't insert another row.  I'd consider that a duplicate.

Second, JSON-ifying is not great, but OTOH nothing is great.  Problems with hashing JSON is that the hash depends on non-information-carrying elements of the serialization, for example whitespace and key ordering. That said, I think your implementation should not change the hash as long as nothing else changes inside the `json` lib.  I considered (and I guess slightly prefer) just using Panda's own [`pandas.util.hash_pandas_object`] (https://pandas.pydata.org/docs/reference/api/pandas.util.hash_pandas_object.html) to return one hash value for reach row of the dataframe.  Now, it has similar problems such as depending on column order and on index value.  You can somewhat achieve a similar result, in much less code, using:

```py
    filtered_df["metadata_uuid"] = pd.util.hash_pandas_object(filtered_df.sort_index(axis=1), index=False)
    #optionally wrap this in uuid.uuid5(...)
```

Finally, consider choosing the most appropriate postgresql datatype for `_id`: If it's really a uuid then use `uuid`, if it's an integer (as with hash_pandas_object) then choose `bigint`.

All this is food for thought.  As I said, I am not in love with any solution, so it's also fine if you do not touch this.

Reply:
Thanks for this! I didn't know about `hash_pandas_object`. Despite its own shortcomings, I find that using it nicely streamlines this function where we can now accomplish everything we want with Pandas. 

I'm about to push a commit that implements this change; I also worked in your suggestion to only hash based on the most important columns, and a comment in `alerts_gcs_test` to show how we are testing all of this.